### PR TITLE
ARROW-9106: [Python] Allow specifying CSV file encoding

### PR DESCRIPTION
**Pull request by pitrou**
Also add a C++ InputStream wrapper that transforms data from the stream according to an arbitrary callable.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Should the user guide also be updated? (the changes in `docs/source/cpp/csv.rst ` and `docs/source/python/csv.rst` from https://github.com/apache/arrow/commit/1325d139b7301df40ec0b22a74afd675aebe0d94)

**Comment by pitrou** [comment]
Ah, right, thank you.

File: python/pyarrow/tests/test_io.py [hunk 1]
```diff
@@ -1289,6 +1290,56 @@ def test_compressed_recordbatch_stream(compression):
...
+                          ('utf-16', 'utf-8'),
+                          ('utf-8', 'utf-32-le'),
+                          ('utf-8', 'utf-32-be'),
+                          ])
+def test_transcoding_input_stream(src_encoding, dest_encoding):
+    # All at once
+    check_transcoding(unicode_transcoding_example,
+                      src_encoding, dest_encoding, [1000, 0])
+    # Incremental
+    check_transcoding(unicode_transcoding_example,
```
Comment (pitrou) [inline python/pyarrow/tests/test_io.py:1328]: TODO: should perhaps exercise encoding errors

**Review by wesm — APPROVED** [review]
+1, did you want to do more things in this PR or should I merge?

**Comment by pitrou** [comment]
I'll add a test that encoding errors are propagated correctly (hopefully :-)).

**Comment by pitrou** [comment]
Ok, I added some tests for error propagation. I'm going to merge if CI stays green.

**Comment by wesm** [comment]
Looks good here. Merging

**Pull request closed without merging**